data/tutorials/language/0it_06_hof.md

```diff
@@ -0,0 +1,967 @@
+- : unit = ()
+```
+
+If we wanted to say "hello" to the same person multiple times, we'd just _repeat_ the same line of code.
+
+```ocaml 
+# say_hi "Camel";;
+Hello, Camel!
+- : unit = ()
+
+# say_hi "Camel";;
+Hello, Camel!
+- : unit = ()
+
+# say_hi "Camel";;
+Hello, Camel!
+- : unit = ()
+```
+
+How would we avoid having to repeat this lines every time? One way of doing this would be to write a function to say hi 3 times:
+
+```ocaml 
+# let say_hi_3_times name =
+  say_hi name;
+  say_hi name;
+  say_hi name
+;;
+val say_hi_3_times : string -> unit = <fun>
+```
+
+In this function we can see a few behaviors:
+* It says "hi" to the same name.
+* It repeats it exactly 3 times.
+
+But what would happen if we wanted to say "hi" 2 times? Or 4 or 12 times?
+
+When this happens, it usually means that the function is making certain decision that it shouldn't. In other words, the function **knows something** (like the number of times).
+
+So instead, we will create a function that **let's the caller decide** how many times to say "hi." We do this by requiring a new argument, in this case, `times`:
+
+```ocaml 
+# let rec say_many_hi times name =
+  if times < 1 then ()
+  else begin
+    say_hi name;
+    say_many_hi (times - 1) name
+  end
+;;
+val say_many_hi : int -> string -> unit = <fun>
+```
+
+Much better. Now we can call:
+
+```ocaml
+# say_many_hi 3 "Xavier";;
+Hello, Xavier!
+Hello, Xavier!
+Hello, Xavier!
+- : unit = ()
+
+# say_many_hi 12 "Camel";;
+Hello, Camel!
+Hello, Camel!
+Hello, Camel!
+Hello, Camel!
+Hello, Camel!
+Hello, Camel!
+Hello, Camel!
+Hello, Camel!
+Hello, Camel!
+Hello, Camel!
+Hello, Camel!
+Hello, Camel!
+- : unit = ()
+```
+
+Unfortunately, reusing this _repetition_ behaviour isn't so easy because we have hard-coded our call to `say_hi`. 
+
+To make this reusable, we can **let the caller decide** what our function should do:
+
+```ocaml 
+# let rec repeat times thing_to_do =
+  if times < 1 then ()
+  else begin
+    thing_to_do;
+    repeat (times - 1) thing_to_do
+  end
+;;
+val repeat : int -> 'a -> unit = <fun>
+```
+
+But what should `thing_to_do` be? Our intuition may be that we can call:
+
+```ocaml 
+# repeat 3 (say_hi "Camel");;
+Hello, Camel!
+- : unit = ()
+```
+
+But our program only outputs one salutation:
+
+```
+Hello, Camel!
+```
+
+That is not what we want! We want it to say "hello" to Camel 3 times.
+
+In OCaml, the arguments are evaluated before the function itself, so in this case, we ended up saying "hi" before we even got to the repeat function.
+
+We can let `repeat` call `say_hi` many times by _delaying_ the function's execution by wrapping it with another function, like this:
+
+```diff
+# repeat 3 (
++ fun () ->
+    say_hi "Camel");;
+```
+
+This means we must refactor our `repeat` function a little to _call_ our new function:
+
+```diff
+let rec repeat times thing_to_do =
+  if times < 1 then ()
+  else (
+-   thing_to_do;
++   thing_to_do ();
+    repeat (times - 1) thing_to_do)
+;;
```

Comment:
Same with this one:

`Error: This expression has type int
       This is not a function; it cannot be applied.`